// @fanchenbao/react-native-scroll-indicator/lib/module/index.js
/* eslint-disable react-native/no-inline-styles */
/**
 * @format
 */
import * as React from 'react';
import { ScrollIndicator } from './ScrollIndicator';
import { getDefaultPosition } from './functions';
export const ScrollViewIndicator = props => {
  const {
    position = '',
    horizontal = false,
    persistentScrollbar = false,
    indStyle: {
      width = 5,
      ...indStyle
    } = {},
    scrollViewProps = {}
  } = props;
  return /*#__PURE__*/React.createElement(ScrollIndicator, {
    target: "ScrollView",
    targetProps: scrollViewProps,
    horizontal: horizontal,
    position: getDefaultPosition(horizontal, position),
    persistentScrollbar: persistentScrollbar,
    indStyle: {
      backgroundColor: 'grey',
      width,
      borderRadius: width / 2,
      ...indStyle
    }
  }, props.children);
};
export const FlatListIndicator = props => {
  const {
    flatListProps,
    position = '',
    horizontal = false,
    persistentScrollbar = false,
    indStyle: {
      width = 5,
      ...indStyle
    } = {}
  } = props;
  return /*#__PURE__*/React.createElement(ScrollIndicator, {
    target: "FlatList",
    targetProps: flatListProps,
    horizontal: horizontal,
    position: getDefaultPosition(horizontal, position),
    persistentScrollbar: persistentScrollbar,
    indStyle: {
      backgroundColor: 'grey',
      width,
      borderRadius: width / 2,
      ...indStyle
    }
  });
};
